pkg/deploy/controller/imagechange/factory.go

```diff
@@ -3,82 +3,176 @@ package imagechange
+	// stream stores have synced.
+	StoreSyncedPollPeriod = 100 * time.Millisecond
+	// MaxRetries is the number of times an image stream will be retried before it is dropped
+	// out of the queue.
+	MaxRetries = 5
+)
+
+// NewImageChangeController returns a new ImageChangeController.
+func NewImageChangeController(dcInformer, streamInformer framework.SharedIndexInformer, oc osclient.Interface) *ImageChangeController {
+	c := &ImageChangeController{
+		dn: oc,
+
+		queue: workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+	}
+
+	c.streamStore.Indexer = streamInformer.GetIndexer()
+	streamInformer.AddEventHandler(framework.ResourceEventHandlerFuncs{
+		AddFunc:    c.addImageStream,
+		UpdateFunc: c.updateImageStream,
+	})
+	c.streamStoreSynced = streamInformer.HasSynced
+
+	c.dcStore.Indexer = dcInformer.GetIndexer()
+	dcInformer.AddEventHandler(framework.ResourceEventHandlerFuncs{
+		AddFunc:    c.addDeploymentConfig,
+		UpdateFunc: c.updateDeploymentConfig,
+	})
+	c.dcStoreSynced = dcInformer.HasSynced
+
+	return c
+}
+
+// Run begins watching and syncing.
+func (c *ImageChangeController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	// Wait for the dc store to sync before starting any work in this controller.
+	ready := make(chan struct{})
+	go c.waitForSyncedStore(ready, stopCh)
+	select {
+	case <-ready:
+	case <-stopCh:
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+	glog.Infof("Shutting down image change controller")
+	c.queue.ShutDown()
+}
+
+func (c *ImageChangeController) waitForSyncedStore(ready chan<- struct{}, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	for !c.streamStoreSynced() || !c.dcStoreSynced() {
+		glog.V(4).Infof("Waiting for the image stream and deployment config caches to sync before starting the image change controller workers")
+		select {
+		case <-time.After(StoreSyncedPollPeriod):
+		case <-stopCh:
+			return
+		}
+	}
+	close(ready)
+}
+
+func (c *ImageChangeController) addImageStream(obj interface{}) {
+	stream := obj.(*imageapi.ImageStream)
+	c.enqueueImageStream(stream)
+}
+
+func (c *ImageChangeController) updateImageStream(old, cur interface{}) {
+	// A periodic relist will send update events for all known streams.
+	if kapi.Semantic.DeepEqual(old, cur) {
+		return
+	}
+
+	stream := cur.(*imageapi.ImageStream)
+	c.enqueueImageStream(stream)
+}
+
+func (c *ImageChangeController) addDeploymentConfig(obj interface{}) {
+	dc := obj.(*deployapi.DeploymentConfig)
+	for _, stream := range c.streamStore.GetStreamsForConfig(dc) {
+		glog.V(4).Infof("Reconciling stream %q for config %q\n", stream.Name, dc.Name)
+		c.enqueueImageStream(stream)
+	}
+}
+
+func (c *ImageChangeController) updateDeploymentConfig(old, cur interface{}) {
+	// A periodic relist will send update events for all known streams.
+	if kapi.Semantic.DeepEqual(old, cur) {
```

Comment:
we just picked a change from upstream that cast to concrete types and compared resourceVersion fields... would that be better?